### ISSUE TITLE
Make ChangeDetectorRef#detach() work with most child components

### ISSUE THREAD

**Issue by pauldraper** [angular/angular#39949]
# 🚀 feature request

### Relevant Package

@angular/core

### Description

Detaching views is (sometimes) absolutely necessary for good performance.

```html
<item *ngFor="let item of longList" [value]="item"></item>
```

Whenever a single `item` is invalidated (even for a mousemove event), Angular has to check every single `item`. Unless of course they are detached.

But.....**detach() is practically unusable** because it's very, very poorly supported.

Angular Material is perhaps the best-made Angular library and even it breaks if detaching any parent. Example: https://stackblitz.com/edit/angular-nvcgil (expandable area will not expand)

### Describe the solution you'd like

When I detach from change detection, what I really want is to *not to dirty the parent*, not to change how the children operate.

In other words, I want to make the component a "root" for change detection.

AFAIK there is no mechanism to do that currently.

### Describe alternatives you've considered

* Not detaching. The causes poor performance.
* Fixed libraries for detached change detection. But I'm 100% certain the Angular team will refuse my PRs to Angular Material.
* Not using libraries.

It's unfortunate having to choose between using libraries and having performance.

**Comment by pauldraper** [comment]
Workaround:

```ts
const flags = Symbol("FLAGS");

export function patchView(changeDetector: ChangeDetectorRef) {
  const detectChanges = () => changeDetector.detectChanges();

  const view = <any>changeDetector;
  const lView = view._cdRefInjectingView || view._lView;

  lView[flags] = lView[2];
  Object.defineProperty(lView, 2, {
    get() {
      return this[flags];
    },
    set(v: any) {
      this[flags] = v;
      if (v & 64) {
        Promise.resolve().then(detectChanges);
      }
    }
  });
}
```

Then in the component constructor, call

```ts
patchView(changeDetector);
```

https://stackblitz.com/edit/angular-dn8j6s

---

Seems kinda iffy though....would appreciate alternative suggestions for detach w/ third-party libs.

**Comment by JiaLiPassion** [comment]
I think the current behavior is correct.

If my understanding of your angular material sample https://stackblitz.com/edit/angular-nvcgil is correct,
 in the expansion click event handler, now the change detection is from `root`, so `AppComponent` is detected, and nothing will be updated, what you want is to trigger change detection not from `root`, but from the `current component` (the current means the component contains the event handler) ?

**Comment by pauldraper** [comment]
Yes, you understand correctly.

This is a feature request for a detach() that is actually usable.

Right now, almost nothing the ecosystem supports it, and it seems impossible (except earlier hack mucking with View), to make them work with it.

**Comment by sod** [comment]
We already have pretty good performance with `changeDetection: ChangeDetectionStrategy.OnPush,` (this already makes your component kinda a new root component - especially if it has no inputs) plus a `trackBy` function on the loop.

I wonder what your additional perf gains are on top of that with detach.

**Comment by pauldraper** [comment]
I presented it already in my earlier comment:

```
<item *ngFor="let item of longList" [value]="item"></item>
```

Angular has to run through a long list every single time something changes in one of them. Maybe I'll put together benchmarks, but it's pretty obvious what the issue is (and why someone added `detach()` in the first place).
